### PR TITLE
FBW preformance enhancements

### PR DESCRIPTION
**Pull request by gmarcos87**
These changes change the following behavior: 
- Before doing the scan sort the networks by type and channel. 
- At the moment of connecting it checks if the channel and mode is the same that is configured.
- If it is the same it avoids the configuration change
- If it is not the same make the necessary changes for the connection

Doing this saves a lot of time and improves the performance of the utility.

It also adds functions like reduce, flatTable, sortBy and splitBy to functools.

### CODE REVIEW

**Comment by gmarcos87** [comment]
Some of the changes are discussed in [this post](https://mesheando.net/tutoriales/03-lua-closure/) in Spanish in a very basic way.

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 1]
```diff
@@ -348,17 +357,36 @@ local function printParam(text,campo)
...
+    -- Add lock file
     start_scan_file()
+    -- Clear previus scans
+    clearTmp()
+    -- Set wireless backup
+    backup_wifi_config()
+    -- Get all networks
     local networks = get_networks()
-    ft.map(printParam('FBW MESH - All Networlk ssid','ssid'), networks)
+    -- FIlter only mesh adn ad-hoc
```
Comment (G10h4ck) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: sed 's/adn/and/'

Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: Done in 7d94761

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 2]
```diff
@@ -270,38 +155,31 @@ function clean_lime_config()
...
-local function stop_scan()
+-- Remove scan lock file
+local function end_scan()
     local file = io.open("/tmp/scanning", "w")
     file:write("false")
     file:close()
 end
 
+-- Read scan status
 function check_scan_file()
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: this function is not used, so should be removed

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 3]
```diff
@@ -1,136 +1,107 @@
...
-    return all_networks
+    -- Get all radios
+    local radios = ft.map(utils.extract_prop(".name"), wireless.scandevices())
+    -- Get only 5ghz radios
+    local radios_5ghz = ft.filter(wireless.is5Ghz,  radios)
+    -- Convert radios to phys
+    local phys = ft.map(utils.radio_to_phy, radios_5ghz)
+    -- Scan networks in phys and format result
+    local networks = ft.map(
+        function(phy)
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: this function looks cryptic here, would recommend putting it outside and comment what it does

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    local unpacked = {}
+    for k,v in ipairs(t) do
+        for sk, sv in ipairs(v) do
+            unpacked[#unpacked+1] = sv
+        end
+    end
+    return unpacked
+end
+
+function utils.sortNetworks(networks)
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: rename to sort_by_channel_and_mode so it is clear when you read it

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 5]
```diff
@@ -318,47 +197,66 @@ function check_lock_file()
     return true
 end
 
+-- Remove lock file
 function remove_lock_file()
-    execute("rm /etc/first_run")
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: if we have a function to check it, would also create a function to remove it so it is api symetrical

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 6]
```diff
@@ -318,47 +197,66 @@ function check_lock_file()
...
+            table.insert(result, {
+                ap = string.gsub(ap, "\"", ""),
+                file = file
+            })
+        end
+    end
+    return result
 end
 
+-- Create new network
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: the comment doesn't match the name of the function. please fix

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua [hunk 7]
```diff
@@ -318,47 +197,66 @@ function check_lock_file()
...
     uci_cursor:commit("lime-defaults")
-
-    -- Apply config and reboot
-    execute("rm /etc/config/lime")
+    -- Clean previus lime configuration
     clean_lime_config()
-    execute("/rom/etc/uci-defaults/91_lime-config")
-    execute("rm /etc/first_run")
+    -- Apply new configuration and reboot
+    utils.execute("/rom/etc/uci-defaults/91_lime-config")
```
Comment (nicopace) [inline packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: it is not clear if this is what applies the config or the line two lines below. could you clarify?

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 2 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: Its not used here, but is used by ubus deamon

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 3 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: Changed in favor of **extract_phys_from_radios**

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua (hunk 4 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard/utils.lua]: Change in favor of **sort_by_channel_and_mode**

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 5 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: I don't understand what you're asking for.

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 6 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: I change the comment line: _Apply configuration for a new network ( userd in ubus daemon)_

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 7 above)
Comment (gmarcos87) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: Changed on 8e3e4490dbc7d0a73ff30e764700bab72760e91d. I remove line 232 (Clean...) and move clean_lime_config() after the comment section.

File: packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua (hunk 5 above)
Comment (nicopace) [reply packages/first-boot-wizard/files/usr/lib/lua/firstbootwizard.lua]: my mistake

**Pull request merged**